### ISSUE TITLE
High overhead / poor performance with React SPA and PostHogProvider

### ISSUE THREAD

**Issue by lauri865** [PostHog/posthog-js#1514]
We just spent the last 5 days searching for the root cause of our performance issues in the app, and after optimising everything to no end, we finally realized that the root cause seems to be Posthog. After disabling Posthog, our app became lightning fast in production, and route changes became from sluggish to instant. Click events from lagging to instant.

It worked fine with light-weight pages, but as long as there's something more expensive happening on the page itself (datagrids, charts, etc.), we started hitting many performance issues. As any self-aware development team, we just assumed it was our shitty code at fault, but we assumed completely wrong this time.

It seems that the issue has become much worse in the recent months. For now we decided to turn off Posthog completely, since the performance tax is not worth it for us. Our analytics needs are rather simple in a B2B SaaS setting (whereas performance is one of the key selling points of our application), and thus we're not interested in any advanced instrumentation that could affect UX as a consequence. If anything, we'd be more fine with analytics failing than it affecting the user experience.

It seems like new minor semver versions turn on new features automatically, which is just not acceptable nor predictable. Either we need to stop updating the dependency or stop using the react package altogether.

After some testing, we got to an acceptable performance level, but still with visibly slower navigation and click events vs. not having Posthog enabled. The config is as follows:
```
const options: Partial<PostHogConfig> = {
  api_host: import.meta.env.VITE_POSTHOG_HOST,
  persistence: "memory",
  capture_pageview: false,
  capture_pageleave: false,
  cross_subdomain_cookie: true,
  enable_recording_console_log: false,
  autocapture: false,
  disable_session_recording: true,
  disable_surveys: true,
  enable_heatmaps: false,
  disable_scroll_properties: true,
  disable_web_experiments: true,
  rageclick: false,
  capture_heatmaps: false,
  capture_performance: false,
  capture_dead_clicks: false,
};
```

There's a million flags there and a lack of any documentation – e.g. what are the default values, what they do, etc. Better type hints / descriptions would go a long way here.

In summary, we have no idea which of the many features is actually causing the performance issues, but would love any help with achieving a React setup that has minimal to runtime overhead, and a guidance on how we can avoid automatically opting in to new features in the future, unless we explicitly decide to do so after testing them.

We're on version 1.180.1. React 19 + compiler, react-router v6 SPA.

**Comment by lauri865** [comment]
May be related to #1405

**Comment by playtons** [comment]
AFAIK, to disable Web Vitals, capture_performance: false is insufficient.

The needed one is:

`capture_performance: { web_vitals: false }`

[Source](https://posthog.com/questions/disabling-web-vitals-on-one-domain-when-two-domains-are-tracked-with-the-same-project)

**Comment by kevinfaveri** [comment]
Hello @robbie-c do you know what might be causing the performance impact here? Any option that could be disabled to support this?

**Comment by rafaeelaudibert** [comment]
Hey @lauri865 - and other folk. Can I have access to your website to debug this further? We're using posthog in our own app - duh - and that's a pretty big app with a lot of interactions, and we don't see the problems you're seeing.

Granted, we're not using the same stack (we aren't using React Router or React Compiler [yet]), so I'd like to debug it further. If you don't wanna share it publicly, I'd love it if you could email it to me, rafael@posthog.com. Either your website URL or your public PH key works. Thank you!

**Comment by rafaeelaudibert** [comment]
> AFAIK, to disable Web Vitals, capture_performance: false is insufficient.
> 
> We needed to do:
> 
> `capture_performance: { web_vitals: false }`
> 
> [Source](https://posthog.com/questions/disabling-web-vitals-on-one-domain-when-two-domains-are-tracked-with-the-same-project)

@playtons `capture_performance: false` should do the trick, I'm working on optimizing our docs to reflect that - as suggested by @lauri865 😉

**Comment by rafaeelaudibert** [comment]
https://github.com/PostHog/posthog-js/pull/1684 could help with this

**Comment by marandaneto** [comment]
Closing the issue as a part of large repository cleanup, due to it being inactive and/or outdated. There's a large chance that this issue has been solved since.
Please do not hesitate to ping me if it is still relevant, and I will happily reopen and work on it.
Cheers!

**Comment by aleclarson** [comment]
Reopen this. Posthog affecting page load performance is an important issue.
